[pkg/telemetryquerylanguage] Add support for passing Getter slices to functions

evan-bradley opened the pull request:
**Description:**

This allows TQL functions to receive an array of Getters as a parameter.

**Link to tracking Issue:**

This is a prerequisite for #12476, which will need to concatenate strings and path references to strings and therefore requires the ability to receive an array containing more than one primitive type.

**Testing:**
- [x]  Unit tests

**Documentation:**

There doesn't appear to be a good place to document these changes, but I'd be happy to add docs if there is.

evan-bradley committed:
[pkg/telemetryquerylanguage] Add support for passing Getter slices to functions

This is a prerequisite for #12476.

TylerHelmuth commented:
> There doesn't appear to be a good place to document these changes, but I'd be happy to add docs if there is.

Good callout.  I didn't specify well in the tql README that some Value types could be used as slices in function called.  Can you update the [Invocations section](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/telemetryquerylanguage/tql#invocations) to specify the parameter types that the TQL understands?

evan-bradley committed:
[pkg/telemetryquerylanguage] Fix minor README typo

evan-bradley committed:
[pkg/telemetryquerylanguage] Change package name in error message

The error message should never be printed, but the message should be
changed anyway.

evan-bradley committed:
[pkg/telemetryquerylanguage] Fix minor typo in README

evan-bradley commented:
I've updated the docs to specify what types are supported from a function author's perspective, and to add some context around that I updated the Value types to include slices and added in a description for accessing telemetry data, mainly to discuss the Getter interface.

I also fixed a few typos and put the changes in separate commits. I can break those out into a separate PR, but I figured they weren't too distracting.

TylerHelmuth commented on pkg/telemetryquerylanguage/tql/functions_test.go:
```diff
@@ -200,6 +200,35 @@ func Test_NewFunctionCall(t *testing.T) {
+					{
+						Float: tqltest.Floatp(1.1),
+					},
+					{
```
Good test.  Can you add an arg that is an Invocation.

evan-bradley replied:
Added to the test.

TylerHelmuth commented on pkg/telemetryquerylanguage/tql/README.md:
```diff
@@ -17,15 +17,41 @@ Invocations represent a function call. Invocations are made up of 2 parts
+- [Paths](#paths).
+- [Literals](#literals).
+- [Enums](#enums).
+- [Slices](#slices).
```
This isn't technically correct, since the language cannot differentiate between arguments that are individual versus combined into a slice as part of the function definition.  The rest of the change is good, but I think Slices should be removed.

evan-bradley replied:
I somehow missed that, I've removed them as a Value type.

TylerHelmuth commented on pkg/telemetryquerylanguage/tql/README.md:
```diff
@@ -17,15 +17,41 @@ Invocations represent a function call. Invocations are made up of 2 parts
 
+#### Invocation paramaters
+
+The TQL will use reflection to determine parameter types when parsing an invocation within a statement.
```
```suggestion
The TQL will use reflection to determine parameter types when parsing an invocation within a statement.  When interpreting slice parameter types, the TQL will attempt to build the slice from all remaining Values in the Invocation's arguments.  As a result, function implementations of Invocations may only contain one slice argument and it must be the last argument in the function definition.  See [function syntax guidelines](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/processing.md#function-syntax) for more details.
```

evan-bradley replied:
Thanks, I've added this.

TylerHelmuth commented on pkg/telemetryquerylanguage/tql/README.md:
```diff
@@ -91,6 +116,29 @@ Operators determine how the two Values are compared.  The valid operators are:
 - Equal (`==`). Equal (`==`) checks if the left and right Values are equal, using Go's `==` operator.
 - Not Equal (`!=`).  Not Equal (`!=`) checks if the left and right Values are not equal, using Go's `!=` operator.
 
+#### Slices
```
I think we can remove this Slices section since slices are not part of the grammar itself yet but are a function implementation concern.

evan-bradley replied:
Good call, I've removed it.

TylerHelmuth reviewed (commented):
/cc @kentquirk

kentquirk reviewed (approved):
This is very useful. Thank you.

TylerHelmuth committed:
Merge branch 'main' into issue-12476-tql

TylerHelmuth commented:
@open-telemetry/collector-contrib-maintainer can we merge this?

Commit:
Merge branch 'main' into issue-12476-tql
